design/proposals/ksonnet-lib-simple-constructors.md

```diff
@@ -0,0 +1,109 @@
+
+## Motivation
+
+* `ksonnet-lib` includes constructors for popular resources. In a few cases, these constructors are incomplete or include items which may not be required. The incompleteness and inconsistencies are a barrier to adoption because developers do not know how to get started in a productive way.
+* `ksonnet-lib` should be a direct translation of the OpenAPI swagger.json.
+
+## Goal
+
+* Developers should be able to construct objects using the builder pattern rather than constructors.
+* Upstream projects should have a clean slate to build abstraction upon.
+
+## Non Goals
+
+The following are related but not the goals of this specific proposal
+
+* Create upstream projects that implement constructors (e.g. v1 Deployment)
+
+## Proposal
+
+The change would be introducing a new constructor for all resources.
+
+```jsonnet
+group:: {
+  v1:: {
+    local apiVersion = {apiVersion: "group/v1"},
+    kind:: {
+      local kind = {kind: "Kind"},
+      init() :: apiVersion + kind,
+    },
+  },
+}
+```
+
+Many of the existing constructors adhere to this format. This change would make sure all resources have a constructor that only includes the `apiVersion` and `kind`.
+
+## User Experience
+
+### Use Cases
+
+To deprecate the existing the existing custom constructors a new constructor will be introduced, `init()`. Currently, to construct a v1beta2 Deployment, you have to do the following:
+
+```jsonnet
+local params = ...;
+local deployment = k.apps.v1beta2.deployment;
+local container = k.apps.v1beta2.deployment.mixin.spec.template.spec.containersType;
+local containerPort = container.portsType;
+local labels = {app: params.name};
+local targetPort = params.containerPort;
+
+local appDeployment = deployment
+  .new(
+    params.name,
+    params.replicas,
+    container
+      .new(params.name, params.image)
+      .withPorts(containerPort.new(targetPort)),
+    labels
+);
+```
+
+This snippet illustrates a few issues:
+
+* The constructor requires items that aren't required by the spec.
+* The constructor requires a specific order.
+
+Instead of the custom constructor, instead you can construct using init using the following:
+
+```jsonnet
+local params = ...;
+local deployment = k.apps.v1beta2.deployment;
+local containersType = k.apps.v1beta2.deployment.mixin.spec.template.spec.containersType;
+local containerPort = container.portsType;
+local labels = {app: params.name};
+local targetPort = params.containerPort;
+
+local port = containerPort
+  .init()
+  .withContainerPort(targetPort)
+
+local container = containersType
+  .init()
+  .withName(params.name)
+  .withImage(params.image)
+  .withPorts(port)
+
+local appDeployment = deployment
```

Comment:
Hmm, I'm not sure this code actually works? Shouldn't it be something like (I haven't tested this code):

```
deployment
  .init() +
  deployment.mixin.metadata.withName(params.name) +
  deployment.mixin.spec.withReplicas(params.replicas) +
  deployment.mixin.spec.template.spec
    .withContainers(containers)
    .withLabels(labels);
```

You basically have to use a `+` for each "namespace". So everything in `spec.template.spec` can build builder-style constructed altogether, and the `+`'d to the builder-style constructed `spec.template`, `spec`, `metadata`, and so on. I think this gets more obvious when you set more than one property in each, like:

```
  deployment.mixin.metadata
    .withName(params.name)
    .withNamespace(params.namespace) +
  deployment.mixin.spec.template.spec
    .withContainers(containers)
    .withLabels(labels);
```

Reply:
Good catch, I'll update the proposal.